keras_cv/models/segmentation/deeplab_v3_plus/deeplab_v3_plus.py

```diff
@@ -0,0 +1,223 @@
+        num_classes: int, the number of classes for the detection model. Note
+            that the `num_classes` doesn't contain the background class, and the
+            classes from the data should be represented by integers with range
+            [0, `num_classes`).
+        projection_filters: int, number of filters in the
+            convolution layer projecting low-level features from the `backbone`.
+            The default value is set to `48`, as per the
+            [TensorFlow implementation of DeepLab](https://github.com/tensorflow/models/blob/master/research/deeplab/model.py#L676).
+        spatial_pyramid_pooling: (Optional) a `keras.layers.Layer`. Also known
+            as Atrous Spatial Pyramid Pooling (ASPP). Performs spatial pooling
+            on different spatial levels in the pyramid, with dilation. If
+            provided, the feature map from the backbone is passed to it inside
+            the DeepLabV3 Encoder, otherwise
+            `keras_cv.layers.spatial_pyramid.SpatialPyramidPooling` is used.
+        segmentation_head: (Optional) a `keras.layers.Layer`. If provided, the
+            outputs of the DeepLabV3 encoder is passed to this layer and it
+            should predict the segmentation mask based on feature from backbone
+            and feature from decoder, otherwise a similar architecture is used.
+        dropout (Optional) float, The dropout rate applied to the outputs of the
+            decoder in the `segmentation_head`. This parameter is only relevant if
+            `segmentation_head` is `None`, i.e, the default segmentation head is
+            being used.
+
+    Examples:
+    ```python
+    import tensorflow as tf
+    import keras_cv
+
+    images = np.ones(shape=(1, 96, 96, 3))
+    labels = np.zeros(shape=(1, 96, 96, 1))
+    backbone = keras_cv.models.ResNet50V2Backbone(input_shape=[96, 96, 3])
+    model = keras_cv.models.segmentation.DeepLabV3Plus(
+        num_classes=1, backbone=backbone,
+    )
+
+    # Evaluate model
+    model(images)
+
+    # Train model
+    model.compile(
+        optimizer="adam",
+        loss=keras.losses.BinaryCrossentropy(from_logits=False),
+        metrics=["accuracy"],
+    )
+    model.fit(images, labels, epochs=3)
+    ```
+    """
+
+    def __init__(
+        self,
+        backbone,
+        num_classes,
+        projection_filters=48,
+        spatial_pyramid_pooling=None,
+        segmentation_head=None,
+        dropout=0.0,
+        **kwargs,
+    ):
+        if not isinstance(backbone, keras.layers.Layer) or not isinstance(
+            backbone, keras.Model
+        ):
+            raise ValueError(
+                "Argument `backbone` must be a `keras.layers.Layer` instance "
+                f" or `keras.Model`. Received instead "
+                f"backbone={backbone} (of type {type(backbone)})."
+            )
+
+        inputs = backbone.input
+
+        final_backbone_pyramid_output = backbone.get_layer(
```

Comment:
This was done as per [this](https://github.com/keras-team/keras-cv/pull/1831#discussion_r1219765453) request.

Reply:
I think in general we probably _can_ just use the backbone output here.

I brought this up originally because I wasn't clear on whether DeepLabV3Plus always wants to use the highest-available P-level here or if it always wants to use e.g. P4. It seems like it may want the highest P-level always, which I think should mean it always matches up with the backbone output.

Reply:
I don't think the last P-level is *exactly* the same as the output, although it is the same shape. There are often extra normalization and activations (and even extra conv layers!) afterward that we omit for 1:1 comparability with the previous levels (e.g., [ResNetV2](https://github.com/keras-team/keras-cv/blob/b3798d8140605b651ac46f40fe7072ea08175720/keras_cv/models/backbones/resnet_v2/resnet_v2_backbone.py#L143), [MobileNet](https://github.com/keras-team/keras-cv/blob/b3798d8140605b651ac46f40fe7072ea08175720/keras_cv/models/backbones/mobilenet_v3/mobilenet_v3_backbone.py#L155))

Reply:
@jbischof 
I attempted to train with the output of the `backbone` as the features, and it seems that the outputs are significantly worse in that case, compared to the `P5` features.

Here's a comparison: https://wandb.ai/geekyrakshit/deeplabv3-keras-cv/reports/Backbone-Outputs-vs-P5-Features--Vmlldzo0NjQ5NDUw